gateway-util-common/src/main/java/org/apache/knox/gateway/plang/Parser.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.knox.gateway.plang;
+
+import java.io.IOException;
+import java.io.PushbackReader;
+import java.io.StringReader;
+
+public class Parser {
```

Comment:
Should all of this parsing-related code go into the gateway-util-common module or its own module?

Reply:
Good questions. We can extract in into its own module. Although it's only few hundred lines of code. We also need to come up with a good name for the module if we choose to extract it out.

Reply:
I'd like to see it as reusable as possible. I see this AST language approach being applicable across multiple provider types: identity assertion for group mapping, a new authorization provider for richer ACLs/rules, perhaps a new provider type for routing or being used within existing dispatches for the same. I don't know that the initial implementation needs to account for this though as we can refactor later. Putting into gateway-util-common will go a long way though. It will also make it available in a lot of other consumer modules without having to add a new dependency.

Reply:
The parser covers the full syntax, it is unlikely we need to change that unless we want to add new literal syntax for something (e.g.: set/map/vector). So I think it's both stable and reusable. The interpreter is extensible as it allows us to add functions from the outside depending on the use case. We can keep it under gateway-util-common for now and extract later only if needed.